opencog/atoms/distvalue/CTHist.h

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (C) 2019 SingularityNet
+ * All Rights Reserved
+ *
+ * Written by Roman Treutlein
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License v3 as
+ * published by the Free Software Foundation and including the exceptions
+ * at http://opencog.org/wiki/Licenses
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program; if not, write to:
+ * Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#ifndef _OPENCOG_CTHIST_H
+#define _OPENCOG_CTHIST_H
+
+#include <opencog/atoms/distvalue/CoverTree.h>
```

Comment:
I did not read this carefully yet, but it appears that there is nothing in 
the cover-tree code that requires atoms or the atomspace.  As such, 
if it is really generic code, then the `cogutil` repo would be a better place for it.

Reply:
@linas it's because his histograms can be multi-dimensional. I agree that cover-tree should be in `cogutil`, but I actually wonder if this should not go to its own repo, since it is experimental and should remain experimental for a while. @rTreutlein, what do you think?

Reply:
By "this" I mean the whole PR, not just covertree.

Reply:
Sounds reasonable.

Reply:
OK, I've created https://github.com/opencog/distributional-value you can re-create your PRs on that repo.

Reply:
what's multi-dimensionality got to do with nearest neighbors and cover-trees?

Reply:
Well the real issue are dynamic multi dimensional Histograms. Because we don't know beforehand in what range the data will be we can't generate the bins beforehand. So we need to somehow generate the bins on the fly. There aren't really any data structures that can store and update arbitery sized bins without requiring you to rebuild most he structure each time. Wich is why i decided to just store the centers of bins. There still not many structures that can store these points in ordered way that can be easily update but Covertrees can do that better then most as far as i could find. And of course since we only have the centers to work with now it's quite important to find the closest center to some point of interest wich is where nearest neighbor search comes in.
It took quite a while to find a data representation that fit all the requirements. But as the readme doesn't explain that well enough atm ill work on that somemore so it's understandable why covertrees are the right choice in this case.

Reply:
If you can't generate the bins beforehand, then how can you know where the center of the bin is?  But if you do know what the centers of the bins are, then the edges of the bins are just at the half-way-point between two bins!?  I assume that you are still working with rectangular bins, when you go multi-dimensional.

(I fear confusing the issue, but should point out: when the bins aren't rectangular, then you get something called a "Voronoi diagram" in 2D and a "Wigner–Seitz cell" in general N-dimensions. The bin boundaries (cell boundaries) are still given by the half-way points between bin centers.)

It sounds to me like perhaps you are doing, or are attempting to do something called "agglomerative clustering", of assigning new points to the "nearest cluster".  I'm not sure what the best agglomerative clustering algo is; for the langauge learning, I just rolled my own, because it was easy.  Never occurred to me that I needed to somehow optimize the nearest-neighbor search ... Hmm. I will have to ponder what that means for my algos.  (My data is super-high-dimensional, like millions of dimensions, and is also super-sparse, only one in a trillion bins are non-empty.)

Reply:
Ofcourse i also don't know the centers of the bins beforhand. I created them one at a time as the data points come in (each data point becomes a bin center untill i have reach the max number of bins specified). This i could ofcourse also do with normal bins but the issue arises when I have to modify those bins for example splitting a bin or combinign 2 bins. Which i need to do because the first couple of data points that lead to these bins will likely not have created a very good distribution of bins. Doing this is hard without needing to rebuild the whole data structure containing these bins.
Doing this with just centers is easier. And yes i am doing agglomerative clustering. Each data point that get's added to a bin moves the center of the bin a small amount. Which should make the bins converge to a nice distribution over time.
Since only working with centers i have stoped thinking about the bins having a clear boundary so they are certainly not rectangular. I guess i could consider them "Wigner-Seitz cells" but haven't given that much thought yet. What i do if i want to know the expected frequency of a given Value is check the bin center clostes to it and a bin center in the opposite direction to the closest one and then do an interpolation to get my result.
Regarding sparse data my Histograms should be able to deal with that well as they only create bins at locations where we have actually seen data. But haven't really thought about using it for super-high-dimensions or bigger data sets. As i don't expect PLN to generate/use joint distributions that use more then a handfull of dimensions, though this is just my opinion reality might differ.

Reply:
OK, two questions. first, a 1D example: what happens if you decide you want 50 bins max, and then send in points from 1 to 1000 in sequential order (not random order)? Roughly, where do the bin centers end up (guesstimating?)

Why do you need a cover tree to do what you described?